### ISSUE TITLE
[CUDA] illegal memory read on embedding_backward_feature_kernel

### ISSUE THREAD

**Issue by kokol16** [pytorch/pytorch#189668]
### 🐛 Describe the bug

```python
import torch
print("torch:", torch.__version__)
torch.cuda.init()
weight = torch.zeros(10, 0, device='cuda', requires_grad=True)
idx = torch.zeros(3, dtype=torch.long, device='cuda')
out = torch.nn.functional.embedding(idx, weight)
out.sum().backward()
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void at::native::<unnamed>::embedding_backward_feature_kernel<float, float, long>(const T3 *, const T1 *, T1 *, int, long, int)+0x5f0
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x0 is out of bounds
=========     and is 140678598230016 bytes before the nearest allocation at 0x7ff249e00000 of size 2097152 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

**Comment by CodersAcademy006** [comment]
Root cause: `embedding_dense_backward_cuda` launches `embedding_backward_feature_kernel` over a `grad_weight` of shape `{num_weights, grad.size(-1)}`. With zero embedding width the gradient is empty and its `data_ptr` is null, so the kernel does a null `__global__` read (`Address 0x0`), confirmed under `compute-sanitizer` on a Tesla T4.

#189714 returns the correctly-shaped zero gradient before launching any kernel (covers both the feature-kernel and sort-based paths), with a regression test.

**Comment by CodersAcademy006** [comment]
@albanD moving this here as you asked, since the gate you set is the `actionable` label on the issue rather than the PR. This one is already `triaged` with `module: crash`, `module: autograd`, `module: cuda` and `module: embedding`, but it has not been marked `actionable`, so it is stuck one step short.

Concretely: `embedding_dense_backward_cuda` launches `embedding_backward_feature_kernel` over a `grad_weight` of shape `{num_weights, grad.size(-1)}`. When the embedding width is zero that gradient is empty, its `data_ptr` is null, and the kernel performs a null `__global__` read, which is what the reporter's compute-sanitizer trace shows at address `0x0`. The CPU path returns an empty tensor for the same input, so this is CUDA diverging from CPU on a legal input rather than a request for new behaviour. Could you or whoever owns embedding make the `actionable` call, either way? If the answer is that a zero-width embedding is not worth guarding, I am happy to close #189714 and stop here.

**Comment by malfet** [comment]
Duplicate of #195547. This is one instance of a class of missing shape/size checks that lead to out-of-bounds tensor access — here surfaced by compute-sanitizer on a CUDA kernel rather than by ASAN on CPU, but the same missing bounds contract. The class is now tracked under the umbrella issue #195547, which also explains why per-operator TORCH_CHECK PRs are not the fix being sought. Please follow #195547.
